sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVFileFormat.scala

```diff
@@ -169,13 +169,18 @@ private[csv] class CsvOutputWriter(
     context: TaskAttemptContext,
     params: CSVOptions) extends OutputWriter with Logging {
 
-  private val charset = Charset.forName(params.charset)
+  private var univocityGenerator: Option[UnivocityGenerator] = None
```

Comment:
Do we have a race condition below then where multiple generators can be created?

Reply:
We have not observe any race conditions so far. Instances of `UnivocityGenerator` are created per-tasks as well as `OutputStreamWriter`s. They share instances of schema and CSVOptions but we do not modify them while writing. Inside of each `UnivocityGenerator`, we create an instance of `CsvWriter` but I almost absolutely sure they do not share anything internally.

Reply:
I don't mean that it would cause an error, but that it could create many generators and writers that aren't closed. It may not be obvious that it's happening. Unless we know writes will only happen in one thread what about breaking out and synchronizing the get/create part of this method?

Reply:
> ... but that it could create many generators and writers that aren't closed. 

Writers/generators are created inside of tasks: https://github.com/apache/spark/blob/ab1650d2938db4901b8c28df945d6a0691a19d31/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormatWriter.scala#L228-L256 where `dataWriter.commit()` and `dataWriter.abort()` close writers/generators. So, number of not closed generators is less or equal to the size of the task thread pool on executors at any moment.

> Unless we know writes will only happen in one thread ...

According to comments below, this is our assumption: https://github.com/apache/spark/blob/e8167768cfebfdb11acd8e0a06fe34ca43c14648/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormatDataWriter.scala#L33-L37